### DISCUSSION TITLE
Setting persistentVolumeReclaimPolicy to DELETE

### DISCUSSION THREAD

**Discussion by atrzcinski** [kubernetes-csi/csi-driver-smb#570] [Q&A] [answered]
Hi, 

I'm wondering what effect setting persistentVolumeReclaimPolicy of pv from RETAIN to DELETE will be? 
Will the data be deleted or not and why?

**Comment by andyzhangx** [comment, marked as answer]
the mounted folder would be deleted if set as DELETE

**Reply by atrzcinski** [reply]
I tested it and after uninstall the service from kubernetes (with pv) the data from mounted folder is still there

I see in the cs-smb-node* logs that the my-pv is sucessfully onmounted, but the data on mounted folder are not deleted (I checked this on smb mounted folder on my computer):
`I0109 13:23:06.596648       1 nodeserver.go:255] NodeUnstageVolume: CleanupMountPoint on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/my-pv/globalmount with volume unique-volumeid
I0109 13:23:06.596657       1 mount_linux.go:294] Unmounting /var/lib/kubelet/plugins/kubernetes.io/csi/pv/my-pv/globalmount
W0109 13:23:06.627005       1 mount_helper_common.go:133] Warning: "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/my-pv/globalmount" is not a mountpoint, deleting
I0109 13:23:06.627130       1 nodeserver.go:260] NodeUnstageVolume: unmount volume unique-volumeid on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/my-pv/globalmount successfully
I0109 13:23:06.627149       1 utils.go:83] GRPC response: {}`

**Reply by andyzhangx** [reply]
could you follow this guide to get controller logs and check what's wrong with `/DeleteVolume` in the logs? https://github.com/kubernetes-csi/csi-driver-smb/blob/master/docs/csi-debug.md#case1-volume-createdelete-failed

**Reply by atrzcinski** [reply]
I can't see the new logs, there are only old ones.. :( Could I change the log level of this driver to see more?

**Reply by atrzcinski** [reply]
Ok, I found the answer: 

https://cloud.google.com/kubernetes-engine/docs/how-to/access-smb-volume
The warning note: "Warning: The Delete [reclaim policy](https://kubernetes.io/docs/tasks/administer-cluster/change-pv-reclaim-policy/) of a PersistentVolume or a StorageClass is not supported by the open source SMB CSI Driver for Kubernetes."

so the mounted folder won't be deleted

**Reply by andyzhangx** [reply]
delete reclaim policy is supported in storage class at least, it was added long time ago: https://github.com/kubernetes-csi/csi-driver-smb/pull/259, while for PV, it's tricky since when you use a PV, it means you bring your own smb server dir, and then you want the driver to delete that dir.

**Reply by andyzhangx** [reply]
if you add annotations `pv.kubernetes.io/provisioned-by: smb.csi.k8s.io`
```
  annotations:
    pv.kubernetes.io/provisioned-by: smb.csi.k8s.io
```
into following PV config, it should work:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: smb.csi.k8s.io
  name: pv-smb
spec:
```

**Reply by atrzcinski** [reply]
added PV config with annotations on test evn, it doesn't change anything, the folder still contains folders and files after uninstall with policy DELETE, no "/DeleteVolume" in the logs of smb-controller, smb-node-*, maybe it work for storage class, but for PV not.

**Reply by andyzhangx** [reply]
then you should have the `/DeleteVolume` logs in driver controller:
https://github.com/kubernetes-csi/csi-driver-smb/blob/master/docs/csi-debug.md#case1-volume-createdelete-failed

**Reply by atrzcinski** [reply]
there are only startup logs, no new logs provided after mount / unmount

`> cat csi-smb-controller.log 
W0112 12:06:24.625992       1 main.go:62] nodeid is empty
I0112 12:06:24.631401       1 main.go:92] set up prometheus server on 0.0.0.0:29644
I0112 12:06:24.631704       1 smb.go:91] 
DRIVER INFORMATION:
-------------------
Build Date: "2022-08-20T07:34:36Z"
Compiler: gc
Driver Name: smb.csi.k8s.io
Driver Version: v1.9.0
Git Commit: ""
Go Version: go1.17.3
Platform: linux/amd64

Streaming logs below:
I0112 12:06:24.631842       1 mount_linux.go:208] Detected OS without systemd
I0112 12:06:24.631860       1 driver.go:93] Enabling controller service capability: CREATE_DELETE_VOLUME
I0112 12:06:24.631867       1 driver.go:93] Enabling controller service capability: SINGLE_NODE_MULTI_WRITER
I0112 12:06:24.631874       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_WRITER
I0112 12:06:24.631883       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_READER_ONLY
I0112 12:06:24.631887       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_SINGLE_WRITER
I0112 12:06:24.631892       1 driver.go:112] Enabling volume access mode: SINGLE_NODE_MULTI_WRITER
I0112 12:06:24.631896       1 driver.go:112] Enabling volume access mode: MULTI_NODE_READER_ONLY
I0112 12:06:24.631910       1 driver.go:112] Enabling volume access mode: MULTI_NODE_SINGLE_WRITER
I0112 12:06:24.631928       1 driver.go:112] Enabling volume access mode: MULTI_NODE_MULTI_WRITER
I0112 12:06:24.631933       1 driver.go:103] Enabling node service capability: STAGE_UNSTAGE_VOLUME
I0112 12:06:24.631939       1 driver.go:103] Enabling node service capability: SINGLE_NODE_MULTI_WRITER
I0112 12:06:24.631944       1 driver.go:103] Enabling node service capability: VOLUME_MOUNT_GROUP
I0112 12:06:24.631949       1 driver.go:103] Enabling node service capability: GET_VOLUME_STATS
I0112 12:06:24.632154       1 server.go:118] Listening for connections on address: &net.UnixAddr{Name:"//csi/csi.sock", Net:"unix"}
I0112 12:06:24.699739       1 utils.go:76] GRPC call: /csi.v1.Identity/GetPluginInfo
I0112 12:06:24.699768       1 utils.go:77] GRPC request: {}
I0112 12:06:24.703908       1 utils.go:83] GRPC response: {"name":"smb.csi.k8s.io","vendor_version":"v1.9.0"}
I0112 12:06:25.386019       1 utils.go:76] GRPC call: /csi.v1.Identity/GetPluginInfo
I0112 12:06:25.386055       1 utils.go:77] GRPC request: {}
I0112 12:06:25.386147       1 utils.go:83] GRPC response: {"name":"smb.csi.k8s.io","vendor_version":"v1.9.0"}
I0112 12:06:25.387656       1 utils.go:76] GRPC call: /csi.v1.Identity/GetPluginCapabilities
I0112 12:06:25.387675       1 utils.go:77] GRPC request: {}
I0112 12:06:25.387837       1 utils.go:83] GRPC response: {"capabilities":[{"Type":{"Service":{"type":1}}}]}
I0112 12:06:25.388829       1 utils.go:76] GRPC call: /csi.v1.Controller/ControllerGetCapabilities
I0112 12:06:25.388846       1 utils.go:77] GRPC request: {}
I0112 12:06:25.388883       1 utils.go:83] GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":13}}}]}
> `

**Reply by andyzhangx** [reply]
then what's the pv status? `kubectl describe pv pv-name`

**Reply by atrzcinski** [reply]
after install (mount) the pv is in Bound status, after uninstall:

`Error from server (NotFound): persistentvolumes "my-pv" not found`

**Reply by andyzhangx** [reply]
try following example directly:

```
kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/csi-driver-smb/master/deploy/example/pvc-smb.yaml
kubectl apply -f https://github.com/kubernetes-csi/csi-driver-smb/blob/master/deploy/example/pvc-smb-static.yaml
# wait about 10s to make sure pvc is in Bound state
kubectl get pvc pvc-smb

# delete pvc and then pv will also be deleted
kubectl delete -f https://github.com/kubernetes-csi/csi-driver-smb/blob/master/deploy/example/pvc-smb-static.yaml
```

 - in csi driver controller logs, it would how have following logs which means DeleteVolume happens, and since subDir is not in volumeID, it's skippped
```
I0113 09:42:04.021816       1 utils.go:76] GRPC call: /csi.v1.Controller/DeleteVolume
I0113 09:42:04.021851       1 utils.go:77] GRPC request: {"volume_id":"smb-server.default.svc.cluster.local/share##"}
I0113 09:42:04.021997       1 controllerserver.go:174] DeleteVolume(smb-server.default.svc.cluster.local/share##) does not delete subdirectory
```

**Comment by andyzhangx** [comment]
this doc PR would fix the issue: https://github.com/kubernetes-csi/csi-driver-smb/pull/572
